Support pytest without -s

Currently if pytest is executed without -s, then many tests fail due to e.g.:

```
___________________________________________________________________________________ test_runner_reloads_server_on_different_gid ____________________________________________________________________________________

    def test_runner_reloads_server_on_different_gid():
        # Given the server has been started with real gid 1
        # And the decorated function is executed with real gid 2
        # Then the server should be reloaded, and the decorated function executed
        @cli_factory(current_test_name())
        def runner():
            def inner():
                output_file.write_text(
                    f'{os.getpid()} {os.getppid()}', encoding='utf-8')
                return 0
            return inner

        with isolated_filesystem() as path:
            output_file = path / 'test.txt'

            with contained_children():
                with patch('os.getgid', lambda: 1):
>                   assert runner() == 0

tests/cli_wrapper/test_cli_wrapper.py:1114:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
quicken/lib/_decorator.py:81: in wrapper
    user_data=user_data,
quicken/lib/_lib.py:141: in server_runner_wrapper
    response = client.send(req)
quicken/lib/_client.py:18: in send
    self._client.send(request)
/home/chris/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/connection.py:206: in send
    self._send_bytes(_ForkingPickler.dumps(obj))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'multiprocessing.reduction.ForkingPickler'>, obj = <quicken.lib._protocol.Request object at 0x7f392dd83860>, protocol = None

    @classmethod
    def dumps(cls, obj, protocol=None):
        buf = io.BytesIO()
        #cls(buf, protocol).dump(obj)
>       cls(buf, protocol).dump(obj)
E       TypeError: cannot serialize '_io.FileIO' object
```

The reason is because `sys.stdout` and `sys.stderr` are replaced with `EncodedFile` by pytest, and these are not recognized as io.TextIOWrapper objects. The reason it shows up as _io.FileIO is because that is the failed member of the input objects (EncodedFile).

Options:

1. Support transfer of FileIO objects (or the containing EncodedFile)
   * Should be easier to implement
2. Wrap objects with `io.TextIOWrapper`
   * Could also fix issues like pytest-dev/pytest#4843

Since our use case is sending these entities over an fd, and there would be no way to distinguish them on the other side except type, 1 is probably the best option.

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -105,7 +105,7 @@ jobs:
     displayName: Run tests
     inputs:
       scriptPath: .ap/scripts/python.py
-      arguments: -m poetry run pytest -s -ra --junitxml=junit/test-results.xml tests
+      arguments: -m poetry run pytest -ra --junitxml=junit/test-results.xml tests
 
   - task: PublishTestResults@2
     inputs:

--- a/quicken/lib/_multiprocessing_asyncio.py
+++ b/quicken/lib/_multiprocessing_asyncio.py
@@ -100,6 +100,7 @@ class AsyncConnectionAdapter:
             async with self._disconnect_cv:
                 self._disconnect_cv.notify_all()
         self._detach_from_event_loop()
+        # XXX: Should include __cause__ or context.
         self._read_queue.put_nowait(ConnectionClose())
         self._connected = False
         self._loop.create_task(signal_disconnect())

--- a/quicken/lib/_multiprocessing_reduction.py
+++ b/quicken/lib/_multiprocessing_reduction.py
@@ -1,10 +1,14 @@
+"""Functions for translating between low-level structures
+for passing over multiprocessing Connections.
+"""
 from __future__ import annotations
 
 import os
+import pickle
 import threading
 
-from io import TextIOWrapper
-from multiprocessing.reduction import recv_handle, register, send_handle
+from io import BytesIO, FileIO, TextIOWrapper
+from multiprocessing.reduction import recv_handle, send_handle
 
 from ._imports import multiprocessing_connection
 from ._signal import blocked_signals, signal_range
@@ -15,6 +19,39 @@ if MYPY_CHECK_RUNNING:
     from typing import TextIO
 
 
+class Pickler(pickle.Pickler):
+    """Custom pickler class to avoid changing the global multiprocessing.reduction
+    pickler.
+    """
+    _dispatch_table = {}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.dispatch_table = self._dispatch_table
+
+    @classmethod
+    def dumps(cls, obj):
+        buf = BytesIO()
+        cls(buf).dump(obj)
+        buf.seek(0)
+        return buf.read()
+
+    @classmethod
+    def register(cls, cls2, reduce):
+        cls._dispatch_table[cls2] = reduce
+
+
+def dumps(obj):
+    return Pickler.dumps(obj)
+
+
+loads = pickle.loads
+
+
+def register(cls, reduce):
+    return Pickler.register(cls, reduce)
+
+
 _fd_sharing_base_path_fd = None
 
 
@@ -23,7 +60,14 @@ def set_fd_sharing_base_path_fd(fd: int):
     _fd_sharing_base_path_fd = fd
 
 
-def reduce_textio(obj: TextIO):
+class Fd:
+    """Tag for reduce_fd.
+    """
+    def __init__(self, fd):
+        self.fd = fd
+
+
+def reduce_fd(obj: Fd):
     """Simpler version of multiprocessing.resource_sharer._ResourceSharer
     that:
 
@@ -32,12 +76,7 @@ def reduce_textio(obj: TextIO):
     * doesn't require random unix socket (so when we import multiprocessing.connection
       we can stub out tempfile, saving 5ms).
     """
-    # Picklable object that contains a callback id to be used by the
-    # receiving process.
-    if obj.readable() == obj.writable():
-        raise ValueError(
-            'TextIO object must be either readable or writable, but not both.')
-    fd = obj.fileno()
+    fd = obj.fd
     name = f'{os.getpid()}-{fd}'
 
     with chdir(_fd_sharing_base_path_fd):
@@ -59,22 +98,58 @@ def reduce_textio(obj: TextIO):
 
     t = threading.Thread(target=target)
     t.daemon = True
+    # Blocking signals for any utility threads gives the main app the final say
+    # w.r.t. signal disposition.
     with blocked_signals(signal_range):
-        # Inherits blocked signals.
         t.start()
 
-    return rebuild_textio, (path, obj.readable(), obj.writable())
+    return rebuild_fd, (path,)
 
 
-def rebuild_textio(path: str, readable: bool, _writable: bool) -> TextIO:
+def rebuild_fd(path: str) -> Fd:
     # UNIX socket path name is limited to 108 characters, so cd to the directory
     # and refer to it as a relative path.
     with chdir(os.path.dirname(path)):
         with multiprocessing_connection.Client(os.path.basename(path)) as c:
             c.send(os.getpid())
-            fd = recv_handle(c)
+            return Fd(recv_handle(c))
+
+
+register(Fd, reduce_fd)
+
+
+def reduce_textio(obj: TextIO):
+    if obj.readable() == obj.writable():
+        raise ValueError(
+            'TextIO object must be either readable or writable, but not both.')
+    fd = Fd(obj.fileno())
+    return rebuild_textio, (fd, obj.readable(), obj.writable(), obj.encoding)
+
+
+def rebuild_textio(fd: Fd, readable: bool, _writable: bool, encoding: str) -> TextIO:
     flags = 'r' if readable else 'w'
-    return open(fd, flags)
+    return open(fd.fd, flags, encoding=encoding)
 
 
 register(TextIOWrapper, reduce_textio)
+
+
+def reduce_fileio(obj: FileIO):
+    # Needed for pytest, which sets stdout and stderr to objects that contain
+    # unbuffered temporary files.
+    # File descriptors duplicated across processes preserve offset and other
+    # properties so they should remain synchronized.
+    fd = Fd(obj.fileno())
+    mode = obj.mode
+    return rebuild_fileio, (fd, mode)
+
+
+def rebuild_fileio(fd: Fd, mode: str):
+    # open returns a FileIO object when binary + buffering=0
+    if 'b' not in mode:
+        mode += 'b'
+    # Opening an existing file descriptor with 'w' does not truncate the file.
+    return open(fd.fd, mode, buffering=0)
+
+
+register(FileIO, reduce_fileio)

--- a/quicken/lib/_protocol.py
+++ b/quicken/lib/_protocol.py
@@ -7,7 +7,7 @@ import os
 import sys
 
 # Registers TextIOWrapper handler.
-from . import _multiprocessing_reduction
+from ._multiprocessing_reduction import dumps, loads
 from ._typing import MYPY_CHECK_RUNNING
 
 if MYPY_CHECK_RUNNING:
@@ -40,6 +40,25 @@ class StdStreams:
         self.stdin = stdin
         self.stdout = stdout
         self.stderr = stderr
+
+    def __getstate__(self):
+        """We implement a custom __getstate__ because Python has a default __reduce_ex__
+        that attempts to call __getstate__ on the members of our class.
+
+        When that happens and when using pytest's built-in std stream capturing, this
+        invokes _io.FileIO.__getstate__ which unconditionally raises an exception
+        ('cannot serialize _io.FileIO') in Python 3.7 and below.
+        """
+        return (
+            dumps(self.stdin),
+            dumps(self.stdout),
+            dumps(self.stderr),
+        )
+
+    def __setstate__(self, state):
+        self.stdin = loads(state[0])
+        self.stdout = loads(state[1])
+        self.stderr = loads(state[2])
 
 
 class ProcessState:

--- a/quicken/lib/_server.py
+++ b/quicken/lib/_server.py
@@ -306,7 +306,7 @@ class ProcessConnectionHandler(ConnectionHandler):
         async def accept_request():
             try:
                 request: Request = await connection.recv()
-            except ConnectionClose:
+            except ConnectionClose as e:
                 logger.debug('Connection closed')
             except ConnectionResetError:
                 logger.debug('Connection reset')
@@ -318,8 +318,8 @@ class ProcessConnectionHandler(ConnectionHandler):
 
             # This occurs when we have disconnected from the client so cancel
             # any pending responses and kill the child process.
-            logger.debug('Killing child process')
             if process:
+                logger.debug('Killing child process')
                 try:
                     process.kill()
                 except ProcessLookupError:

--- a/tests/test__multiprocessing.py
+++ b/tests/test__multiprocessing.py
@@ -6,8 +6,14 @@ from pathlib import Path
 
 import pytest
 
+# noinspection PyProtectedMember
 from quicken.lib._multiprocessing import run_in_process
-from quicken.lib._multiprocessing_reduction import set_fd_sharing_base_path_fd
+# noinspection PyProtectedMember
+from quicken.lib._multiprocessing_reduction import (
+    dumps,
+    loads,
+    set_fd_sharing_base_path_fd,
+)
 
 from .utils import isolated_filesystem
 from .utils.process import contained_children
@@ -35,8 +41,10 @@ def test_function_is_executed_in_separate_process():
 #  multiprocessing.
 def test_function_exception_is_reraised():
     message = 'example message'
+
     def runner():
         raise RuntimeError(message)
+
     with pytest.raises(RuntimeError) as exc_info:
         run_in_process(target=runner)
     assert message == exc_info.value.args[0]
@@ -44,24 +52,30 @@ def test_function_exception_is_reraised():
 
 def test_function_return_value_is_returned():
     value = 123
+
     def runner():
         return value
+
     result = run_in_process(target=runner)
     assert result == value
 
 
 def test_function_gets_args():
     arg = '123'
+
     def runner(s):
         return s
+
     result = run_in_process(target=runner, args=(arg,))
     assert result == arg
 
 
 def test_function_timeout_works():
     timeout = 0.5
+
     def runner():
         time.sleep(timeout * 2)
+
     with pytest.raises(TimeoutError):
         run_in_process(target=runner, timeout=timeout)
 
@@ -83,10 +97,11 @@ def test_function_detach_works():
 @pytest.mark.timeout(5)
 def test_textiowrapper_serialization_works():
     shared_text = 'hello\n'
+
     def acceptor():
         listener = multiprocessing.connection.Listener('socket')
         conn = listener.accept()
-        obj = conn.recv()
+        obj = loads(conn.recv())
         obj.write(shared_text)
         obj.flush()
         obj.close()
@@ -103,7 +118,7 @@ def test_textiowrapper_serialization_works():
         r_obj = os.fdopen(r, closefd=False)
         w_obj = os.fdopen(w, 'w', closefd=False)
         c = multiprocessing.connection.Client('socket')
-        c.send(w_obj)
+        c.send(dumps(w_obj))
         text = r_obj.readline()
         assert text == shared_text
         os.close(r)

--- a/tests/utils/pytest.py
+++ b/tests/utils/pytest.py
@@ -1,9 +1,14 @@
+from __future__ import annotations
+
 import os
 import re
 import sys
 
 import pytest
 
+from _pytest.capture import DontReadFromInput, EncodedFile
+
+from quicken.lib._multiprocessing_reduction import register
 
 windows_only = pytest.mark.skipif(
     not sys.platform.startswith('win'), reason='Windows only')
@@ -23,3 +28,20 @@ def current_test_name():
     if not m:
         raise RuntimeError(f'Could not extract test name from {name}')
     return m.group('name')
+
+
+def reduce_encodedfile(obj: EncodedFile):
+    """Register handler for the replaced stdout/stderr that pytest
+    uses.
+    """
+    return EncodedFile, (obj.buffer, obj.encoding)
+
+
+register(EncodedFile, reduce_encodedfile)
+
+
+def reduce_dontreadfrominput(obj: DontReadFromInput):
+    return DontReadFromInput, tuple()
+
+
+register(DontReadFromInput, reduce_dontreadfrominput)

--- a/tests/utils/watch.py
+++ b/tests/utils/watch.py
@@ -34,8 +34,8 @@ def wait_for_delete(path: BoundPath, timeout: float = 5) -> bool:
 def _wait_for(action: Action, timeout: float = 5) -> bool:
     """
     Args:
-        path
-        timeout
+        action:
+        timeout:
     """
     def action_check():
         return action.path.exists() == action.present

